src/java/htsjdk/samtools/util/CollectionUtil.java

```diff
@@ -101,24 +97,38 @@ public void appendAll(final K k, final Collection<? extends V> v) {
 
         private void initializeKeyIfUninitialized(final K k) {
             if (!this.containsKey(k))
-                this.put(k, new LinkedList<V>());
+                this.put(k, new LinkedList<>());
         }
     }
 
-    /** 
+    /**
      * Partitions a collection into groups based on a characteristics of that group.  Partitions are embodied in a map, whose keys are the
      * value of that characteristic, and the values are the partition of elements whose characteristic evaluate to that key.
      */
-    public static <K, V> Map<K,Collection<V>> partition(final Collection<V> collection,  final Partitioner<V, K> p) {
-        final MultiMap<K, V> partitionToValues = new MultiMap<K, V>();
+    @Deprecated //use the version below instead!
+    public static <K, V> Map<K, Collection<V>> partition(final Collection<V> collection, final Partitioner<V, K> p) {
+        final MultiMap<K, V> partitionToValues = new MultiMap<>();
         for (final V entry : collection) {
             partitionToValues.append(p.getPartition(entry), entry);
         }
         return partitionToValues;
     }
+
     public static abstract class Partitioner<V, K> {
         public abstract K getPartition(final V v);
     }
+
+    /** 
+     * Partitions a collection into groups based on a characteristics of that group.  Partitions are embodied in a map, whose keys are the
+     * value of that characteristic, and the values are the partition of elements whose characteristic evaluate to that key.
+     */
+    public static <K, V> Map<K, Collection<V>> partition(final Collection<V> collection, final Function<? super V, ? extends K> keyer) {
```

Comment:
1) good point. I guess I could change the usages of partition to use groupBy instead.
2) I did this initially, but then I felt bad for breaking some possible use of a downstream tool that uses htsjdk...so I did it this way to avoid breaking the API...